qiskit/circuit/oracle/oracle.py

```diff
@@ -0,0 +1,131 @@
+except Exception:  # pylint: disable=broad-except
+    HAS_TWEEDLEDUM = False
+from qiskit import QuantumCircuit, QuantumRegister
+from qiskit.circuit.gate import Gate
+from qiskit.exceptions import QiskitError
+from qiskit.transpiler.oracle_synthesis.utils import tweedledum2qiskit
+from .oracle_visitor import OracleVisitor
+
+
+class Oracle(Gate):
+    """An oracle object represents an oracle function and its logic network."""
+
+    def __init__(self, source, name=None):
+        """Creates a ``Oracle`` from Python source code in ``source``. The code should be
+        a single function with types.
+
+        Args:
+            source (str): Python code with type hints.
+            name (str): Optional. Default: "*oracle*". Oracle name.
+        Raises:
+            ImportError: If tweedledum is not installed.
+            QiskitError: If source is not a string.
+        """
+        if not isinstance(source, str):
+            raise QiskitError('Oracle needs a source code as a string.')
+        if not HAS_TWEEDLEDUM:
+            raise ImportError("To use the oracle compiler, tweedledum "
+                              "must be installed. To install tweedledum run "
+                              '"pip install tweedledum".')
+        self._ast = ast.parse(source)
+        self._network = None
+        self._scopes = None
+        self._args = None
+        super().__init__(name or '*oracle*',
+                         num_qubits=sum([qreg.size for qreg in self.qregs]),
+                         params=[])
+
+    def compile(self):
+        """Parses and creates the logical circuit"""
+        _oracle_visitor = OracleVisitor()
+        _oracle_visitor.visit(self._ast)
+        self._network = _oracle_visitor._network
+        self._scopes = _oracle_visitor.scopes
+        self._args = _oracle_visitor.args
+        self.name = _oracle_visitor.name
+
+    @property
+    def network(self):
+        """Returns the logical network"""
+        if self._network is None:
+            self.compile()
+        return self._network
+
+    @property
+    def scopes(self):
+        """Returns the scope dict"""
+        if self._scopes is None:
+            self.compile()
+        return self._scopes
+
+    @property
+    def args(self):
+        """Returns the oracle arguments"""
+        if self._args is None:
+            self.compile()
+        return self._args
+
+    @property
+    def types(self):
+        """Dumps a list of scopes with their variables and types.
+        Returns:
+            list(dict): A list of scopes as dicts, where key is the variable name and
+            value is its type.
+        """
+        ret = []
+        for scope in self.scopes:
+            ret.append({k: v[0] for k, v in scope.items()})
+        return ret
+
+    def simulate(self):
+        """Runs ``tweedledum.simulate`` on the logic network."""
+        return simulate(self._network)
+
+    def synth(self, arg_regs=False) -> QuantumCircuit:
```

Comment:
`arg_regs` is kind of hard to understand. maybe use `registerless=True` - if set False, then makes registers with the argument names.

Reply:
The default is True?